assembly/runtime.ts

```diff
@@ -801,26 +841,37 @@ export class RootContext extends BaseContext {
    * @param timeout_milliseconds Timeout for the request, in milliseconds.
    * @param cb Callback to be invoked when the request is complete.
    */
-  httpCall(cluster: string, headers: Headers, body: ArrayBuffer, trailers: Headers,
-    timeout_milliseconds: u32, cb: HttpCallback): WasmResultValues {
-
+  httpCall(cluster: string, headers: Headers, body: ArrayBuffer, trailers: Headers, timeout_milliseconds: u32, origin_context_id: u32, cb: (origin_context_id:u32, headers: u32,  body_size: usize, trailers: u32) => void): WasmResultValues {
     let buffer = String.UTF8.encode(cluster);
     let header_pairs = serializeHeaders(headers);
     let trailer_pairs = serializeHeaders(trailers);
-    let token = globalU32Ref;
+    let token = new Reference<u32>();
+    this.http_calls_.set(token.data, new HttpCallback(origin_context_id, cb));
+    log(LogLevelValues.debug, "http_calls_: " + this.http_calls_.size.toString());
+    log(LogLevelValues.debug, "set token: " + token.data.toString() + " on " + this.context_id.toString());
     let result = imports.proxy_http_call(changetype<usize>(buffer), buffer.byteLength, changetype<usize>(header_pairs), header_pairs.byteLength, changetype<usize>(body), body.byteLength, changetype<usize>(trailer_pairs), trailer_pairs.byteLength, timeout_milliseconds, token.ptr());
-    if (result == WasmResultValues.Ok) {
-      this.http_calls_.set(token.data, cb);
-    }
+    log(LogLevelValues.debug, "result: " + result.toString());
     return result;
   }
+  
   onHttpCallResponse(token: u32, headers: u32, body_size: u32, trailers: u32): void {
-    if (this.http_calls_.has(token)) {
-      let callback = this.http_calls_.get(token);
-      this.http_calls_.delete(token);
-      callback.call();
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + " : onHttpCallResponse(token: " + token.toString() + ", headers:" + headers.toString() + ", body_size:" + body_size.toString() + ", trailers:" + trailers.toString() + ")");
+    log(LogLevelValues.debug, "http_calls_: " + this.http_calls_.size.toString());
+    log(LogLevelValues.debug, "get token: " + token.toString() + " from " + this.context_id.toString());
+    log(LogLevelValues.debug, "context_map: " + context_map.keys().join(", "));
+    const readToken = get_buffer_bytes(BufferTypeValues.HttpCallResponseBody, 0, token);
```

Comment:
Also, if you see here, I'm hardcoding the token value because I can't seem to find a way to read the value. I think the token:u32 value being passed here is not the actual token value. I think it is the size of it in the stream. I don't really know how to read it. Could you please help me that?

Reply:
are you using token as the length aregument to get bytes?
the token is essentially a handle. when you do an `proxy_http_call` you get it using a pointer.
on `onHttpCallResponse` you get the token back by value. i.e. the `token` parameter in `onHttpCallResponse` is the same as `token.data` in `httpCall`

the idea is that you call `proxy_http_call`, get a `token`, save it internally, and then use it to correlate the response when it comes i this function.

Reply:
Ok. I think I got this now. I will make the change and test it out.